### DISCUSSION TITLE
HiLetGo ESP-WROOM-32D: No DAC's Available

### DISCUSSION THREAD

**Discussion by Lasakro** [arendst/Tasmota#18549] [Q&A]
Hello,
1) I've flashed the board using https://tasmota.github.io/install/
2) With Tasmotizer retrieved a DHCP derived IP
3) Browsing in it's been configured as a ESP32-DevKit
4) Both GPIO21 & 22 have no DAC Output option.

What am I missing?

**Comment by barbudor** [comment]
DAC output are available only through Berry language

https://tasmota.github.io/docs/Berry/#gpio-module

**Reply by Lasakro** [reply]
I appreciate your assistance with setting my path. I created a dac.be file loaded on boot to configure the port and now can command an output with console changing the output voltage. Still reading about Berry- MQTT to figure out how to command that output from Home Assistant. Nothing is standing out, even from the Berry Cookbook. Will the DAC port be "published" to MQTT? To me since the value will be read from MQTT it should be something different like poll from MQTT but not finding anything yet.

**Reply by sfromis** [reply]
No, there is no automatic publishing around DAC, everything happening must be controlled from your Berry code, including handling of MQTT. As DAC is output, there is just not anything to publish. What you may need is a command; via `tasmota.add_cmd` you can register your Berry code [as a command](https://tasmota.github.io/docs/Berry/#functions-to-create-custom-tasmota-command), which can also be used via MQTT; a good fit for Home Assistant too.

**Reply by pillepalle127** [reply]
I am having the same target. Is there a script available by now?

**Reply by s-hadinger** [reply]
It would be easy to create in Berry a new Tasmota command like "DACOutput" to set the value, and trigger it from MQTT. Can you give us more details on what you are trying to achieve?

**Reply by sfromis** [reply]
Rudimentary sample code for a command like `DACVoltage2 3.0`:
```
tasmota.add_cmd('DACVoltage', def (cmd,idx,payload,data) gpio.dac_voltage(idx+24, int(real(payload)*1000)) tasmota.resp_cmnd_done() end)
```
In general, I'd suggest some use case specific Berry coding around the raw setting of voltage.

**Reply by s-hadinger** [reply]
Nice 1-liner !

**Reply by pillepalle127** [reply]
Tanks! I'll have to try it asap.
